# Data analysis of RNA-seq and TRAP-seq
This repository contains the code for analysis of RNA-seq and TRAP-seq data of [Imai et al. (2023) *Simultaneous measurement of nascent transcriptome and translatome using 4-thiouridine metabolic RNA labeling and translating ribosome affinity purification*](https://www.biorxiv.org/content/10.1101/2023.02.15.525786v1).  
</br>

## Raw sequence data
The raw deep sequencing data are deposited at DDBJ (DRA015587). The data contains two FASTQ files per sample due to paired-end sequencing (e.g., `DRR438599_1.fastq` sequenced from the 5' side and `DRR438599_2.fastq` sequenced from the 3' side, respectively). We only used `_1.fastq` for the analysis. In the `_2.fastq`, the sequencing starts with the poly(T) stretch on the 3' side and sequences through the homopolymer stretch, reducing the quality of sequence reads.  
</br>

## Data Analysis with SLAM-DUNK
The raw sequence data were analyzed using SLAM-DUNK v 0.4.2 ([Neuman T. et al., 2019](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-2849-7)), a pipeline for SLAMseq data analysis, with the default parameters. The analysis environment was built with the following code in Docker.
```
docker pull tobneu/slamdunk  # latest version in 2022
```
The following code is an example for analysis by SLAM-DUNK. The sequence data were aligned on genome-wide 3' UTR sequences generated based on the human genome sequence (GRCh38.p13) and BED file containing coordinates for 3â€™ UTRs (`gencode.v41.costomized.canonical.bed`) generated from `gencode.v41.annotation.gff3` using `generate_costomized_BED.ipynb`. `gencode.v41.costomized.canonical.bed` and `generate_costomized_BED.ipynb` are in `/slamdunk/bed`. `alleyoop utrrates` was used to calculate the T>C conversion ratio. 

```
slamdunk map --trim-5p 12 --max-polya 4 --topn 100 --threads 16 --skip-sam --reference <reference fasta> --outputDir slamdunk/map <fastq file>

slamdunk filter --threads 16 --outputDir slamdunk/filter slamdunk/map/*.bam

alleyoop utrrates --outputDir slamdunk/utrrates --reference <reference fasta> --bed <bed file> -l 100 -t 12 slamdunk/filter/*.bam

slamdunk snp --outputDir slamdunk/snp --threads 16 --reference <reference fasta> slamdunk/filter/*.bam

slamdunk count --threads 16 --reference <reference fasta> --bed <bed file> --outputDir slamdunk/count --snp-directory slamdunk/snp slamdunk/filter/*.bam
```  

## Data Analysis with R
The differential gene expression analysis was performed using DESeq2 v 1.3.8 ([Love M. I. et al., 2014](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8)). Differentially transcribed genes and differential translation efficiency genes (FDR < 0.05) were identified following the deltaTE method ([Chothani S. et al., 2019](https://currentprotocols.onlinelibrary.wiley.com/doi/10.1002/cpmb.108)), with read counts generated by SLAM-DUNK from cytosolic RNA-seq and P-TRAP-seq as input. The input txt files were generated using `/ipynb/generateTXT_for_DESeq2.ipynb` and `/ipynb/generateTXT_for_deltaTE.ipynb`. The R scripts for DESeq2 and deltaTE are located in `/rscript`.  
</br>

## Creating Figures with Jupyter Notebook
All figures were created using Jupyter Notebook files in `/ipynb`. 
