{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0fe16e98",
   "metadata": {},
   "source": [
    "## gencode.v39.annotation.gff3\n",
    "gff[\"attributes\"]に様々な情報が格納されている。各情報は\";\"で区切られているので`split(\";\")`で情報を抽出できる。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f9bc9321",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import gffpandas.gffpandas as gffpd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6a4581c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3238846\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000223972.5;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seq_id  source        type  start    end score strand phase  \\\n",
       "0   chr1  HAVANA        gene  11869  14409     .      +     .   \n",
       "1   chr1  HAVANA  transcript  11869  14409     .      +     .   \n",
       "2   chr1  HAVANA        exon  11869  12227     .      +     .   \n",
       "3   chr1  HAVANA        exon  12613  12721     .      +     .   \n",
       "4   chr1  HAVANA        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...  \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000223972.5;...  \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gencode.v39.annotation.gff3 was downloaded from (https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_39/gencode.v39.annotation.gff3.gz)\n",
    "\n",
    "annotation = gffpd.read_gff3(\"gencode.v39.annotation.gff3\")\n",
    "gff = annotation.df\n",
    "\n",
    "print(len(gff))\n",
    "\n",
    "gff.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82e9951a",
   "metadata": {},
   "source": [
    "タンパク質をコードする遺伝子（`gene_type=\"protein_coding\"`）については、GFFファイル中に3'UTRの情報(`type=\"three_prime_RNA\"`)が含まれているため、その情報を抽出すればOK。一方、lncRNA（`gene_type=lncRNA`）については、GFFファイル中に3'UTRの情報が含まれていないため、代わりに遺伝子全長の情報（`type=\"gene\"`）を抽出する。"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10900f6e",
   "metadata": {},
   "source": [
    "必要な情報は`gene_id`, `gene_name`, `gene_type` の3種類。これらの情報を抽出しする関数を作成"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6a462e27",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_attributes(df):\n",
    "    attributes = df.split(\";\")\n",
    "    output_name = [s for s in attributes if \"gene_name=\" in s][0].replace(\"gene_name=\", \"\")\n",
    "    output_id = [s for s in attributes if \"gene_id=\" in s][0].replace(\"gene_id=\", \"\")\n",
    "    output_type = [s for s in attributes if \"gene_type=\" in s][0].replace(\"gene_type=\", \"\")\n",
    "    return output_name + \"|\" + output_id + \"|\" + output_type\n",
    "\n",
    "def get_gene_type(df):\n",
    "    attributes = df.split(\";\")\n",
    "    output_type = [s for s in attributes if \"gene_type=\" in s][0].replace(\"gene_type=\", \"\")\n",
    "    return output_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cfc5d2c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff[\"gene\"] = gff[\"attributes\"].apply(get_attributes)\n",
    "gff[\"gene_type\"] = gff[\"attributes\"].apply(get_gene_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b6c4b030",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff = gff[((gff[\"gene_type\"] == \"protein_coding\") & (gff[\"type\"] == \"three_prime_UTR\")) | \n",
    "               ((gff[\"gene_type\"] == \"lncRNA\") & (gff[\"type\"] == \"gene\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fba1c1b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>70009</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000641515.2;Parent=ENST000006415...</td>\n",
       "      <td>OR4F5|ENSG00000186092.7|protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>971</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>944154</td>\n",
       "      <td>944574</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000616016.5;Parent=ENST000006160...</td>\n",
       "      <td>SAMD11|ENSG00000187634.13|protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1004</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>944154</td>\n",
       "      <td>944574</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000618323.5;Parent=ENST000006183...</td>\n",
       "      <td>SAMD11|ENSG00000187634.13|protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1050</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>944154</td>\n",
       "      <td>944574</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000342066.8;Parent=ENST000003420...</td>\n",
       "      <td>SAMD11|ENSG00000187634.13|protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1205</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>944154</td>\n",
       "      <td>944575</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000341065.8;Parent=ENST000003410...</td>\n",
       "      <td>SAMD11|ENSG00000187634.13|protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     seq_id  source             type   start     end score strand phase  \\\n",
       "68     chr1  HAVANA  three_prime_UTR   70009   71585     .      +     .   \n",
       "971    chr1  HAVANA  three_prime_UTR  944154  944574     .      +     .   \n",
       "1004   chr1  HAVANA  three_prime_UTR  944154  944574     .      +     .   \n",
       "1050   chr1  HAVANA  three_prime_UTR  944154  944574     .      +     .   \n",
       "1205   chr1  HAVANA  three_prime_UTR  944154  944575     .      +     .   \n",
       "\n",
       "                                             attributes  \\\n",
       "68    ID=UTR3:ENST00000641515.2;Parent=ENST000006415...   \n",
       "971   ID=UTR3:ENST00000616016.5;Parent=ENST000006160...   \n",
       "1004  ID=UTR3:ENST00000618323.5;Parent=ENST000006183...   \n",
       "1050  ID=UTR3:ENST00000342066.8;Parent=ENST000003420...   \n",
       "1205  ID=UTR3:ENST00000341065.8;Parent=ENST000003410...   \n",
       "\n",
       "                                          gene       gene_type  \n",
       "68      OR4F5|ENSG00000186092.7|protein_coding  protein_coding  \n",
       "971   SAMD11|ENSG00000187634.13|protein_coding  protein_coding  \n",
       "1004  SAMD11|ENSG00000187634.13|protein_coding  protein_coding  \n",
       "1050  SAMD11|ENSG00000187634.13|protein_coding  protein_coding  \n",
       "1205  SAMD11|ENSG00000187634.13|protein_coding  protein_coding  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff = gff[gff[\"gene_type\"] == \"protein_coding\"]\n",
    "gff.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5f45c092",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>70009</td>\n",
       "      <td>71585</td>\n",
       "      <td>OR4F5|ENSG00000186092.7|protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944154</td>\n",
       "      <td>944574</td>\n",
       "      <td>SAMD11|ENSG00000187634.13|protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944154</td>\n",
       "      <td>944575</td>\n",
       "      <td>SAMD11|ENSG00000187634.13|protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944154</td>\n",
       "      <td>944259</td>\n",
       "      <td>SAMD11|ENSG00000187634.13|protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944203</td>\n",
       "      <td>944693</td>\n",
       "      <td>NOC2L|ENSG00000188976.11|protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seq_id   start     end                                      gene  score  \\\n",
       "0   chr1   70009   71585    OR4F5|ENSG00000186092.7|protein_coding      0   \n",
       "1   chr1  944154  944574  SAMD11|ENSG00000187634.13|protein_coding      0   \n",
       "2   chr1  944154  944575  SAMD11|ENSG00000187634.13|protein_coding      0   \n",
       "3   chr1  944154  944259  SAMD11|ENSG00000187634.13|protein_coding      0   \n",
       "4   chr1  944203  944693   NOC2L|ENSG00000188976.11|protein_coding      0   \n",
       "\n",
       "  strand  \n",
       "0      +  \n",
       "1      +  \n",
       "2      +  \n",
       "3      +  \n",
       "4      -  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff = gff[[\"seq_id\", \"start\", \"end\", \"gene\", \"score\", \"strand\"]].drop_duplicates().reset_index(drop=True)\n",
    "gff[\"score\"] = 0\n",
    "\n",
    "gff.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77ff7e2b",
   "metadata": {},
   "source": [
    "各遺伝子毎にかぶっている領域をマージする"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "652e6cc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = gff[\"gene\"].unique()\n",
    "output_df = pd.DataFrame(columns=[\"chr\", \"start\", \"end\", \"gene\", \"score\", \"strand\"])\n",
    "\n",
    "for gene in gene_list:\n",
    "    tmp_df = gff[gff[\"gene\"] == gene]    \n",
    "    tmp_chr = tmp_df.iloc[0, 0]\n",
    "    tmp_start = tmp_df[\"start\"].min()\n",
    "    tmp_end = tmp_df[\"end\"].max()\n",
    "    tmp_strand = tmp_df.iloc[0, 5]\n",
    "    \n",
    "    add_row = pd.DataFrame([tmp_chr, tmp_start, tmp_end, gene, 0, tmp_strand], index=output_df.columns).T\n",
    "    \n",
    "    output_df = pd.concat([output_df, add_row])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c3c944c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_df.to_csv(\"gencode.v39.costomized.bed\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f97522dd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
