{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0fe16e98",
   "metadata": {},
   "source": [
    "## gencode.v41.annotation.gff3 の読み込み\n",
    "gff[\"attributes\"]に様々な情報が格納されている。各情報は\";\"で区切られているので`split(\";\")`で情報を抽出できる。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "f9bc9321",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gffpandas.gffpandas as gffpd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "6a4581c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3373604\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000223972.5;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  source        type  start    end score strand phase  \\\n",
       "0  chr1  HAVANA        gene  11869  14409     .      +     .   \n",
       "1  chr1  HAVANA  transcript  11869  14409     .      +     .   \n",
       "2  chr1  HAVANA        exon  11869  12227     .      +     .   \n",
       "3  chr1  HAVANA        exon  12613  12721     .      +     .   \n",
       "4  chr1  HAVANA        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...  \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000223972.5;...  \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  "
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation = gffpd.read_gff3(\"gencode.v41.annotation.gff3\")\n",
    "gff = annotation.df\n",
    "gff = gff.rename(columns={\"seq_id\":\"chr\"}) # 染色体の列名を chr に変更\n",
    "\n",
    "print(len(gff))\n",
    "\n",
    "gff.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82e9951a",
   "metadata": {},
   "source": [
    "タンパク質をコードする遺伝子（`gene_type=\"protein_coding\"`）については、GFFファイル中に3'UTRの情報(`type=\"three_prime_RNA\"`)が含まれており、その情報を抽出すればOK。一方、lncRNA（`gene_type=lncRNA`）については、GFFファイル中に3'UTRの情報が含まれていないため、代わりに遺伝子全長の情報（`type=\"transcript\"`）を抽出する。GFFファイルには1つの遺伝子あたり複数の転写産物が登録されているが、その中で最も代表的なもの (`Ensembl_canonical`) がEnsemblのアルゴリズムによって選出されている。この**canonical**な転写産物のみを抽出する。\n",
    "### Ensembl Canonical transcript\n",
    "The Ensembl Canonical transcript is a single, representative transcript identified at every locus. For accurate analysis, we recommend that more than one transcripts at a locus may need to be considered, however, we designate a single Ensembl Canonical transcript per locus to provide consistency when only one transcript is required e.g. for initial display on the Ensembl (or other) website and for use in genome-wide calculations e.g. the Ensembl gene tree analysis.\n",
    "\n",
    "For protein-coding genes, we aim to identify the transcript that, on balance, has the highest coverage of conserved exons, highest expression, longest coding sequence and is represented in other key resources, such as NCBI and UniProt. To identify this transcript, we consider, where available, evidence of functional potential (such as evolutionary conservation of a coding region, transcript expression levels), transcript length and evidence from other resources (such as concordance with the APPRIS1 ’principal isoform’ and with the UniProt/Swiss-Prot ‘canonical isoform’)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80ac72cc",
   "metadata": {},
   "source": [
    "まず`gene_type`を抽出する関数を用意。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "966e16ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_type(df):\n",
    "    attributes = df.split(\";\")\n",
    "    output_type = [s for s in attributes if \"gene_type=\" in s][0].replace(\"gene_type=\", \"\")\n",
    "    return output_type"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "973b7abf",
   "metadata": {},
   "source": [
    "### (1) タンパク質をコードしている遺伝子のみを抽出\n",
    "`gene_type` が `protein_coding` かつ `Ensembl_canonical` なものを抽出。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "744bd93f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff[\"gene_type\"] = gff[\"attributes\"].apply(get_gene_type)\n",
    "gff = gff[gff[\"gene_type\"] == \"protein_coding\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "c51cd0ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 2988184\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000186092.7;gene_id=ENSG00000186092.7...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000641515.2;Parent=ENSG00000186092.7;...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65419</td>\n",
       "      <td>65433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000641515.2:1;Parent=ENST0000064...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65520</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000641515.2:2;Parent=ENST0000064...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65565</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENST00000641515.2;Parent=ENST0000064151...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr  source        type  start    end score strand phase  \\\n",
       "57  chr1  HAVANA        gene  65419  71585     .      +     .   \n",
       "58  chr1  HAVANA  transcript  65419  71585     .      +     .   \n",
       "59  chr1  HAVANA        exon  65419  65433     .      +     .   \n",
       "60  chr1  HAVANA        exon  65520  65573     .      +     .   \n",
       "61  chr1  HAVANA         CDS  65565  65573     .      +     0   \n",
       "\n",
       "                                           attributes       gene_type  \n",
       "57  ID=ENSG00000186092.7;gene_id=ENSG00000186092.7...  protein_coding  \n",
       "58  ID=ENST00000641515.2;Parent=ENSG00000186092.7;...  protein_coding  \n",
       "59  ID=exon:ENST00000641515.2:1;Parent=ENST0000064...  protein_coding  \n",
       "60  ID=exon:ENST00000641515.2:2;Parent=ENST0000064...  protein_coding  \n",
       "61  ID=CDS:ENST00000641515.2;Parent=ENST0000064151...  protein_coding  "
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Number of rows: {len(gff)}\")\n",
    "gff.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "5831d1fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 518563\n"
     ]
    }
   ],
   "source": [
    "gff = gff[gff[\"attributes\"].str.contains(\"Ensembl_canonical\")]\n",
    "\n",
    "print(f\"Number of rows: {len(gff)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bf6cec7",
   "metadata": {},
   "source": [
    "3'UTR (type=`three_prime_RNA`) の情報のみを抽出する。Sourceは `HAVANA` を利用。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "214c75cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 22243\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>70009</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000641515.2;Parent=ENST000006415...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1008</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>944154</td>\n",
       "      <td>944574</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000616016.5;Parent=ENST000006160...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1318</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>944203</td>\n",
       "      <td>944693</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000327044.7;Parent=ENST000003270...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1386</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>965192</td>\n",
       "      <td>965719</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000338591.8;Parent=ENST000003385...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1445</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>974576</td>\n",
       "      <td>975865</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000379410.8;Parent=ENST000003794...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr  source             type   start     end score strand phase  \\\n",
       "68    chr1  HAVANA  three_prime_UTR   70009   71585     .      +     .   \n",
       "1008  chr1  HAVANA  three_prime_UTR  944154  944574     .      +     .   \n",
       "1318  chr1  HAVANA  three_prime_UTR  944203  944693     .      -     .   \n",
       "1386  chr1  HAVANA  three_prime_UTR  965192  965719     .      +     .   \n",
       "1445  chr1  HAVANA  three_prime_UTR  974576  975865     .      +     .   \n",
       "\n",
       "                                             attributes       gene_type  \n",
       "68    ID=UTR3:ENST00000641515.2;Parent=ENST000006415...  protein_coding  \n",
       "1008  ID=UTR3:ENST00000616016.5;Parent=ENST000006160...  protein_coding  \n",
       "1318  ID=UTR3:ENST00000327044.7;Parent=ENST000003270...  protein_coding  \n",
       "1386  ID=UTR3:ENST00000338591.8;Parent=ENST000003385...  protein_coding  \n",
       "1445  ID=UTR3:ENST00000379410.8;Parent=ENST000003794...  protein_coding  "
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff = gff[(gff[\"type\"] == \"three_prime_UTR\") & (gff[\"source\"] == \"HAVANA\")]\n",
    "print(f\"Number of rows: {len(gff)}\")\n",
    "gff.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0656b332",
   "metadata": {},
   "source": [
    "BEDファイルに記載しておく情報として `gene_name` と `gene_id` の2つの情報があればいい。これらの情報を抽出しする関数を作成。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "75f28a9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_attributes(df):\n",
    "    attributes = df.split(\";\")\n",
    "    gene_name = [s for s in attributes if \"gene_name=\" in s][0].replace(\"gene_name=\", \"\")\n",
    "    gene_id = [s for s in attributes if \"gene_id=\" in s][0].replace(\"gene_id=\", \"\")\n",
    "    return gene_name + \"|\" + gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "36be40ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff[\"gene\"] = gff[\"attributes\"].apply(get_attributes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "fe38f2d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 22243\n",
      "Unique rows: 19487\n"
     ]
    }
   ],
   "source": [
    "rows = len(gff)\n",
    "unique_rows = len(gff[\"gene\"].unique())\n",
    "\n",
    "print(f\"Number of rows: {rows}\")\n",
    "print(f\"Unique rows: {unique_rows}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0563bc64",
   "metadata": {},
   "source": [
    "ユニークな遺伝子名が19,487個あるのに対し行数が多い。被っているものがあるはず。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "b502fb8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7813</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1787322</td>\n",
       "      <td>1787330</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000378609.9;Parent=ENST000003786...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GNB1|ENSG00000078369.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7814</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1785286</td>\n",
       "      <td>1787053</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000378609.9;Parent=ENST000003786...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GNB1|ENSG00000078369.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13647</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>4774500</td>\n",
       "      <td>4774558</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000378191.5;Parent=ENST000003781...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AJAP1|ENSG00000196581.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13648</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>4782545</td>\n",
       "      <td>4792534</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000378191.5;Parent=ENST000003781...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>AJAP1|ENSG00000196581.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15128</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>6106234</td>\n",
       "      <td>6106279</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000262450.8;Parent=ENST000002624...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CHD5|ENSG00000116254.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3372770</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>25037992</td>\n",
       "      <td>25038097</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000382287.5;Parent=ENST000003822...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C|ENSG00000185894.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3372771</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>25031317</td>\n",
       "      <td>25031441</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000382287.5;Parent=ENST000003822...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C|ENSG00000185894.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3372772</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>25030901</td>\n",
       "      <td>25031222</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000382287.5;Parent=ENST000003822...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C|ENSG00000185894.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3373214</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>57128472</td>\n",
       "      <td>57130289</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000286448.12_PAR_Y;Parent=ENST00...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>VAMP7|ENSG00000124333.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3373318</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>57196930</td>\n",
       "      <td>57197337</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000244174.11_PAR_Y;Parent=ENST00...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>IL9R|ENSG00000124334.18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3919 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr  source             type     start       end score strand phase  \\\n",
       "7813     chr1  HAVANA  three_prime_UTR   1787322   1787330     .      -     .   \n",
       "7814     chr1  HAVANA  three_prime_UTR   1785286   1787053     .      -     .   \n",
       "13647    chr1  HAVANA  three_prime_UTR   4774500   4774558     .      +     .   \n",
       "13648    chr1  HAVANA  three_prime_UTR   4782545   4792534     .      +     .   \n",
       "15128    chr1  HAVANA  three_prime_UTR   6106234   6106279     .      -     .   \n",
       "...       ...     ...              ...       ...       ...   ...    ...   ...   \n",
       "3372770  chrY  HAVANA  three_prime_UTR  25037992  25038097     .      -     .   \n",
       "3372771  chrY  HAVANA  three_prime_UTR  25031317  25031441     .      -     .   \n",
       "3372772  chrY  HAVANA  three_prime_UTR  25030901  25031222     .      -     .   \n",
       "3373214  chrY  HAVANA  three_prime_UTR  57128472  57130289     .      +     .   \n",
       "3373318  chrY  HAVANA  three_prime_UTR  57196930  57197337     .      +     .   \n",
       "\n",
       "                                                attributes       gene_type  \\\n",
       "7813     ID=UTR3:ENST00000378609.9;Parent=ENST000003786...  protein_coding   \n",
       "7814     ID=UTR3:ENST00000378609.9;Parent=ENST000003786...  protein_coding   \n",
       "13647    ID=UTR3:ENST00000378191.5;Parent=ENST000003781...  protein_coding   \n",
       "13648    ID=UTR3:ENST00000378191.5;Parent=ENST000003781...  protein_coding   \n",
       "15128    ID=UTR3:ENST00000262450.8;Parent=ENST000002624...  protein_coding   \n",
       "...                                                    ...             ...   \n",
       "3372770  ID=UTR3:ENST00000382287.5;Parent=ENST000003822...  protein_coding   \n",
       "3372771  ID=UTR3:ENST00000382287.5;Parent=ENST000003822...  protein_coding   \n",
       "3372772  ID=UTR3:ENST00000382287.5;Parent=ENST000003822...  protein_coding   \n",
       "3373214  ID=UTR3:ENST00000286448.12_PAR_Y;Parent=ENST00...  protein_coding   \n",
       "3373318  ID=UTR3:ENST00000244174.11_PAR_Y;Parent=ENST00...  protein_coding   \n",
       "\n",
       "                             gene  \n",
       "7813      GNB1|ENSG00000078369.18  \n",
       "7814      GNB1|ENSG00000078369.18  \n",
       "13647    AJAP1|ENSG00000196581.11  \n",
       "13648    AJAP1|ENSG00000196581.11  \n",
       "15128     CHD5|ENSG00000116254.18  \n",
       "...                           ...  \n",
       "3372770   BPY2C|ENSG00000185894.8  \n",
       "3372771   BPY2C|ENSG00000185894.8  \n",
       "3372772   BPY2C|ENSG00000185894.8  \n",
       "3373214  VAMP7|ENSG00000124333.16  \n",
       "3373318   IL9R|ENSG00000124334.18  \n",
       "\n",
       "[3919 rows x 11 columns]"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff[gff[\"gene\"].duplicated(keep=False)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c689712c",
   "metadata": {},
   "source": [
    "3,919個の遺伝子は遺伝子名, 遺伝子IDが被っている。3' UTR の位置が異なっているようだ。例として BPY2C という遺伝子について詳しく見てみる。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "f94b8cbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3372770</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>25037992</td>\n",
       "      <td>25038097</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000382287.5;Parent=ENST000003822...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C|ENSG00000185894.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3372771</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>25031317</td>\n",
       "      <td>25031441</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000382287.5;Parent=ENST000003822...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C|ENSG00000185894.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3372772</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>25030901</td>\n",
       "      <td>25031222</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR3:ENST00000382287.5;Parent=ENST000003822...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>BPY2C|ENSG00000185894.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr  source             type     start       end score strand phase  \\\n",
       "3372770  chrY  HAVANA  three_prime_UTR  25037992  25038097     .      -     .   \n",
       "3372771  chrY  HAVANA  three_prime_UTR  25031317  25031441     .      -     .   \n",
       "3372772  chrY  HAVANA  three_prime_UTR  25030901  25031222     .      -     .   \n",
       "\n",
       "                                                attributes       gene_type  \\\n",
       "3372770  ID=UTR3:ENST00000382287.5;Parent=ENST000003822...  protein_coding   \n",
       "3372771  ID=UTR3:ENST00000382287.5;Parent=ENST000003822...  protein_coding   \n",
       "3372772  ID=UTR3:ENST00000382287.5;Parent=ENST000003822...  protein_coding   \n",
       "\n",
       "                            gene  \n",
       "3372770  BPY2C|ENSG00000185894.8  \n",
       "3372771  BPY2C|ENSG00000185894.8  \n",
       "3372772  BPY2C|ENSG00000185894.8  "
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bpy2c = gff[gff[\"gene\"] == \"BPY2C|ENSG00000185894.8\"]\n",
    "bpy2c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "3a2c2092",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ID=UTR3:ENST00000382287.5;Parent=ENST00000382287.5;gene_id=ENSG00000185894.8;transcript_id=ENST00000382287.5;gene_type=protein_coding;gene_name=BPY2C;transcript_type=protein_coding;transcript_name=BPY2C-201;exon_number=7;exon_id=ENSE00003506422.1;level=2;protein_id=ENSP00000371724.1;transcript_support_level=1;hgnc_id=HGNC:18225;tag=NMD_exception,basic,Ensembl_canonical,MANE_Select,appris_principal_1,CCDS;ccdsid=CCDS44030.1;havana_gene=OTTHUMG00000045199.3;havana_transcript=OTTHUMT00000104944.1',\n",
       "       'ID=UTR3:ENST00000382287.5;Parent=ENST00000382287.5;gene_id=ENSG00000185894.8;transcript_id=ENST00000382287.5;gene_type=protein_coding;gene_name=BPY2C;transcript_type=protein_coding;transcript_name=BPY2C-201;exon_number=8;exon_id=ENSE00001764403.1;level=2;protein_id=ENSP00000371724.1;transcript_support_level=1;hgnc_id=HGNC:18225;tag=NMD_exception,basic,Ensembl_canonical,MANE_Select,appris_principal_1,CCDS;ccdsid=CCDS44030.1;havana_gene=OTTHUMG00000045199.3;havana_transcript=OTTHUMT00000104944.1',\n",
       "       'ID=UTR3:ENST00000382287.5;Parent=ENST00000382287.5;gene_id=ENSG00000185894.8;transcript_id=ENST00000382287.5;gene_type=protein_coding;gene_name=BPY2C;transcript_type=protein_coding;transcript_name=BPY2C-201;exon_number=9;exon_id=ENSE00001740581.1;level=2;protein_id=ENSP00000371724.1;transcript_support_level=1;hgnc_id=HGNC:18225;tag=NMD_exception,basic,Ensembl_canonical,MANE_Select,appris_principal_1,CCDS;ccdsid=CCDS44030.1;havana_gene=OTTHUMG00000045199.3;havana_transcript=OTTHUMT00000104944.1'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bpy2c[\"attributes\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb0e2c4e",
   "metadata": {},
   "source": [
    "この3つは `exon_number` が異なっている。IGVで構造を確認したところ 3' UTR がスプライシングを受けているっぽい。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "66608713",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21206</th>\n",
       "      <td>chrX</td>\n",
       "      <td>299336</td>\n",
       "      <td>303356</td>\n",
       "      <td>PLCXD1|ENSG00000182378.15</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22172</th>\n",
       "      <td>chrY</td>\n",
       "      <td>299336</td>\n",
       "      <td>303356</td>\n",
       "      <td>PLCXD1|ENSG00000182378.15</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr   start     end                       gene  score strand\n",
       "21206  chrX  299336  303356  PLCXD1|ENSG00000182378.15      0      +\n",
       "22172  chrY  299336  303356  PLCXD1|ENSG00000182378.15      0      +"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plcxd1 = gff_dup[gff_dup[\"gene\"] == \"PLCXD1|ENSG00000182378.15\"]\n",
    "plcxd1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7f13dcf",
   "metadata": {},
   "source": [
    "PLCXD1 という遺伝子は X 染色体と Y 染色体の両方に存在しているため重複している。"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3d9b553",
   "metadata": {},
   "source": [
    "同じ染色体内で重複している遺伝子についてはそれぞれ `start` と `end` の最大値と最小値を抽出し、それをマージしたものを3' UTRの範囲とする。"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72219718",
   "metadata": {},
   "source": [
    "まずBEDファイルに必要な情報 (`chr`, `start`, `end`, `gene`, `score`, `strand`) を抽出"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "7eea5a0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>70009</td>\n",
       "      <td>71585</td>\n",
       "      <td>OR4F5|ENSG00000186092.7</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944154</td>\n",
       "      <td>944574</td>\n",
       "      <td>SAMD11|ENSG00000187634.13</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944203</td>\n",
       "      <td>944693</td>\n",
       "      <td>NOC2L|ENSG00000188976.11</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>965192</td>\n",
       "      <td>965719</td>\n",
       "      <td>KLHL17|ENSG00000187961.15</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>974576</td>\n",
       "      <td>975865</td>\n",
       "      <td>PLEKHN1|ENSG00000187583.11</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22238</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25031317</td>\n",
       "      <td>25031441</td>\n",
       "      <td>BPY2C|ENSG00000185894.8</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22239</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25030901</td>\n",
       "      <td>25031222</td>\n",
       "      <td>BPY2C|ENSG00000185894.8</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22240</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25624528</td>\n",
       "      <td>25624902</td>\n",
       "      <td>CDY1|ENSG00000172288.8</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22241</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57128472</td>\n",
       "      <td>57130289</td>\n",
       "      <td>VAMP7|ENSG00000124333.16</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22242</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57196930</td>\n",
       "      <td>57197337</td>\n",
       "      <td>IL9R|ENSG00000124334.18</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22243 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr     start       end                        gene  score strand\n",
       "0      chr1     70009     71585     OR4F5|ENSG00000186092.7      0      +\n",
       "1      chr1    944154    944574   SAMD11|ENSG00000187634.13      0      +\n",
       "2      chr1    944203    944693    NOC2L|ENSG00000188976.11      0      -\n",
       "3      chr1    965192    965719   KLHL17|ENSG00000187961.15      0      +\n",
       "4      chr1    974576    975865  PLEKHN1|ENSG00000187583.11      0      +\n",
       "...     ...       ...       ...                         ...    ...    ...\n",
       "22238  chrY  25031317  25031441     BPY2C|ENSG00000185894.8      0      -\n",
       "22239  chrY  25030901  25031222     BPY2C|ENSG00000185894.8      0      -\n",
       "22240  chrY  25624528  25624902      CDY1|ENSG00000172288.8      0      +\n",
       "22241  chrY  57128472  57130289    VAMP7|ENSG00000124333.16      0      +\n",
       "22242  chrY  57196930  57197337     IL9R|ENSG00000124334.18      0      +\n",
       "\n",
       "[22243 rows x 6 columns]"
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff = gff[[\"chr\", \"start\", \"end\", \"gene\", \"score\", \"strand\"]].drop_duplicates().reset_index(drop=True)\n",
    "gff[\"score\"] = 0\n",
    "\n",
    "gff"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "346c8c7d",
   "metadata": {},
   "source": [
    "重複のある遺伝子とない遺伝子をわける"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "aa7524cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_unique = gff[gff[\"gene\"].duplicated(keep=False) == False]\n",
    "gff_dup = gff[gff[\"gene\"].duplicated(keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "bb1e5b19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>70009</td>\n",
       "      <td>71585</td>\n",
       "      <td>OR4F5|ENSG00000186092.7</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944154</td>\n",
       "      <td>944574</td>\n",
       "      <td>SAMD11|ENSG00000187634.13</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944203</td>\n",
       "      <td>944693</td>\n",
       "      <td>NOC2L|ENSG00000188976.11</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>965192</td>\n",
       "      <td>965719</td>\n",
       "      <td>KLHL17|ENSG00000187961.15</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>974576</td>\n",
       "      <td>975865</td>\n",
       "      <td>PLEKHN1|ENSG00000187583.11</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22219</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22168542</td>\n",
       "      <td>22168819</td>\n",
       "      <td>RBMY1F|ENSG00000169800.14</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22220</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22417604</td>\n",
       "      <td>22417881</td>\n",
       "      <td>RBMY1J|ENSG00000226941.9</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22221</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22514071</td>\n",
       "      <td>22514637</td>\n",
       "      <td>PRY|ENSG00000169789.10</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22229</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24045793</td>\n",
       "      <td>24046065</td>\n",
       "      <td>CDY1B|ENSG00000172352.5</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22240</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25624528</td>\n",
       "      <td>25624902</td>\n",
       "      <td>CDY1|ENSG00000172288.8</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18324 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr     start       end                        gene  score strand\n",
       "0      chr1     70009     71585     OR4F5|ENSG00000186092.7      0      +\n",
       "1      chr1    944154    944574   SAMD11|ENSG00000187634.13      0      +\n",
       "2      chr1    944203    944693    NOC2L|ENSG00000188976.11      0      -\n",
       "3      chr1    965192    965719   KLHL17|ENSG00000187961.15      0      +\n",
       "4      chr1    974576    975865  PLEKHN1|ENSG00000187583.11      0      +\n",
       "...     ...       ...       ...                         ...    ...    ...\n",
       "22219  chrY  22168542  22168819   RBMY1F|ENSG00000169800.14      0      -\n",
       "22220  chrY  22417604  22417881    RBMY1J|ENSG00000226941.9      0      +\n",
       "22221  chrY  22514071  22514637      PRY|ENSG00000169789.10      0      +\n",
       "22229  chrY  24045793  24046065     CDY1B|ENSG00000172352.5      0      -\n",
       "22240  chrY  25624528  25624902      CDY1|ENSG00000172288.8      0      +\n",
       "\n",
       "[18324 rows x 6 columns]"
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "eb1b74c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1787322</td>\n",
       "      <td>1787330</td>\n",
       "      <td>GNB1|ENSG00000078369.18</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1785286</td>\n",
       "      <td>1787053</td>\n",
       "      <td>GNB1|ENSG00000078369.18</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4774500</td>\n",
       "      <td>4774558</td>\n",
       "      <td>AJAP1|ENSG00000196581.11</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4782545</td>\n",
       "      <td>4792534</td>\n",
       "      <td>AJAP1|ENSG00000196581.11</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6106234</td>\n",
       "      <td>6106279</td>\n",
       "      <td>CHD5|ENSG00000116254.18</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22237</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25037992</td>\n",
       "      <td>25038097</td>\n",
       "      <td>BPY2C|ENSG00000185894.8</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22238</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25031317</td>\n",
       "      <td>25031441</td>\n",
       "      <td>BPY2C|ENSG00000185894.8</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22239</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25030901</td>\n",
       "      <td>25031222</td>\n",
       "      <td>BPY2C|ENSG00000185894.8</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22241</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57128472</td>\n",
       "      <td>57130289</td>\n",
       "      <td>VAMP7|ENSG00000124333.16</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22242</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57196930</td>\n",
       "      <td>57197337</td>\n",
       "      <td>IL9R|ENSG00000124334.18</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3919 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr     start       end                      gene  score strand\n",
       "44     chr1   1787322   1787330   GNB1|ENSG00000078369.18      0      -\n",
       "45     chr1   1785286   1787053   GNB1|ENSG00000078369.18      0      -\n",
       "76     chr1   4774500   4774558  AJAP1|ENSG00000196581.11      0      +\n",
       "77     chr1   4782545   4792534  AJAP1|ENSG00000196581.11      0      +\n",
       "80     chr1   6106234   6106279   CHD5|ENSG00000116254.18      0      -\n",
       "...     ...       ...       ...                       ...    ...    ...\n",
       "22237  chrY  25037992  25038097   BPY2C|ENSG00000185894.8      0      -\n",
       "22238  chrY  25031317  25031441   BPY2C|ENSG00000185894.8      0      -\n",
       "22239  chrY  25030901  25031222   BPY2C|ENSG00000185894.8      0      -\n",
       "22241  chrY  57128472  57130289  VAMP7|ENSG00000124333.16      0      +\n",
       "22242  chrY  57196930  57197337   IL9R|ENSG00000124334.18      0      +\n",
       "\n",
       "[3919 rows x 6 columns]"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_dup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "822e4a3d",
   "metadata": {},
   "source": [
    "同じ染色体で重複のある遺伝子について 3' UTR の開始位置の最小値と終了位置の最大値を取得しまとめる。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "4456dbf4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1785286</td>\n",
       "      <td>1787330</td>\n",
       "      <td>GNB1|ENSG00000078369.18</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4774500</td>\n",
       "      <td>4792534</td>\n",
       "      <td>AJAP1|ENSG00000196581.11</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6101787</td>\n",
       "      <td>6106279</td>\n",
       "      <td>CHD5|ENSG00000116254.18</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6159430</td>\n",
       "      <td>6181467</td>\n",
       "      <td>ENSG00000285629|ENSG00000285629.1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10946475</td>\n",
       "      <td>10947662</td>\n",
       "      <td>C1orf127|ENSG00000175262.15</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23285502</td>\n",
       "      <td>23291356</td>\n",
       "      <td>DAZ2|ENSG00000205944.12</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24632011</td>\n",
       "      <td>24639207</td>\n",
       "      <td>BPY2B|ENSG00000183795.8</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24763069</td>\n",
       "      <td>24768933</td>\n",
       "      <td>DAZ3|ENSG00000187191.16</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24901176</td>\n",
       "      <td>24907040</td>\n",
       "      <td>DAZ4|ENSG00000205916.12</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25030901</td>\n",
       "      <td>25038097</td>\n",
       "      <td>BPY2C|ENSG00000185894.8</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1180 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr     start       end                               gene score strand\n",
       "0   chr1   1785286   1787330            GNB1|ENSG00000078369.18     0      -\n",
       "0   chr1   4774500   4792534           AJAP1|ENSG00000196581.11     0      +\n",
       "0   chr1   6101787   6106279            CHD5|ENSG00000116254.18     0      -\n",
       "0   chr1   6159430   6181467  ENSG00000285629|ENSG00000285629.1     0      -\n",
       "0   chr1  10946475  10947662        C1orf127|ENSG00000175262.15     0      -\n",
       "..   ...       ...       ...                                ...   ...    ...\n",
       "0   chrY  23285502  23291356            DAZ2|ENSG00000205944.12     0      +\n",
       "0   chrY  24632011  24639207            BPY2B|ENSG00000183795.8     0      +\n",
       "0   chrY  24763069  24768933            DAZ3|ENSG00000187191.16     0      -\n",
       "0   chrY  24901176  24907040            DAZ4|ENSG00000205916.12     0      +\n",
       "0   chrY  25030901  25038097            BPY2C|ENSG00000185894.8     0      -\n",
       "\n",
       "[1180 rows x 6 columns]"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duplicated_genes = np.array(gff_dup[\"gene\"].unique())\n",
    "gff_dedup = pd.DataFrame(columns=[\"chr\", \"start\", \"end\", \"gene\", \"score\", \"strand\"])\n",
    "\n",
    "for gene in duplicated_genes:\n",
    "    tmp_df = gff_dup[gff_dup[\"gene\"] == gene]   \n",
    "    tmp_df[\"chr\"].unique()   \n",
    "    \n",
    "    if len(tmp_df[\"chr\"].unique()) == 1:\n",
    "        tmp_chr = tmp_df.iloc[0, 0]\n",
    "        tmp_start = tmp_df[\"start\"].min()\n",
    "        tmp_end = tmp_df[\"end\"].max()\n",
    "        tmp_strand = tmp_df.iloc[0, 5]\n",
    "        add_row = pd.DataFrame([tmp_chr, tmp_start, tmp_end, gene, 0, tmp_strand], index=gff_dedup.columns).T\n",
    "        gff_dedup = pd.concat([gff_dedup, add_row])\n",
    "    \n",
    "    else:\n",
    "        gff_dedup = pd.concat([gff_dedup, tmp_df])\n",
    "    \n",
    "gff_dedup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "8d007440",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>70009</td>\n",
       "      <td>71585</td>\n",
       "      <td>OR4F5|ENSG00000186092.7</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944154</td>\n",
       "      <td>944574</td>\n",
       "      <td>SAMD11|ENSG00000187634.13</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944203</td>\n",
       "      <td>944693</td>\n",
       "      <td>NOC2L|ENSG00000188976.11</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>965192</td>\n",
       "      <td>965719</td>\n",
       "      <td>KLHL17|ENSG00000187961.15</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>974576</td>\n",
       "      <td>975865</td>\n",
       "      <td>PLEKHN1|ENSG00000187583.11</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrY</td>\n",
       "      <td>23285502</td>\n",
       "      <td>23291356</td>\n",
       "      <td>DAZ2|ENSG00000205944.12</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24632011</td>\n",
       "      <td>24639207</td>\n",
       "      <td>BPY2B|ENSG00000183795.8</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24763069</td>\n",
       "      <td>24768933</td>\n",
       "      <td>DAZ3|ENSG00000187191.16</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24901176</td>\n",
       "      <td>24907040</td>\n",
       "      <td>DAZ4|ENSG00000205916.12</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25030901</td>\n",
       "      <td>25038097</td>\n",
       "      <td>BPY2C|ENSG00000185894.8</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19504 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr     start       end                        gene score strand\n",
       "0   chr1     70009     71585     OR4F5|ENSG00000186092.7     0      +\n",
       "1   chr1    944154    944574   SAMD11|ENSG00000187634.13     0      +\n",
       "2   chr1    944203    944693    NOC2L|ENSG00000188976.11     0      -\n",
       "3   chr1    965192    965719   KLHL17|ENSG00000187961.15     0      +\n",
       "4   chr1    974576    975865  PLEKHN1|ENSG00000187583.11     0      +\n",
       "..   ...       ...       ...                         ...   ...    ...\n",
       "0   chrY  23285502  23291356     DAZ2|ENSG00000205944.12     0      +\n",
       "0   chrY  24632011  24639207     BPY2B|ENSG00000183795.8     0      +\n",
       "0   chrY  24763069  24768933     DAZ3|ENSG00000187191.16     0      -\n",
       "0   chrY  24901176  24907040     DAZ4|ENSG00000205916.12     0      +\n",
       "0   chrY  25030901  25038097     BPY2C|ENSG00000185894.8     0      -\n",
       "\n",
       "[19504 rows x 6 columns]"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_merged = pd.concat([gff_unique, gff_dedup])\n",
    "gff_merged"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2880e2b6",
   "metadata": {},
   "source": [
    "BEDファイルを保存。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "537d98b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_merged.to_csv(\"gencode.v41.customized.canonical.bed\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56d63561",
   "metadata": {},
   "source": [
    "### lncRNA も含めた BED ファイルを作る"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "aad2e84c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3373604\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000223972.5;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  source        type  start    end score strand phase  \\\n",
       "0  chr1  HAVANA        gene  11869  14409     .      +     .   \n",
       "1  chr1  HAVANA  transcript  11869  14409     .      +     .   \n",
       "2  chr1  HAVANA        exon  11869  12227     .      +     .   \n",
       "3  chr1  HAVANA        exon  12613  12721     .      +     .   \n",
       "4  chr1  HAVANA        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...  \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000223972.5;...  \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  "
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation = gffpd.read_gff3(\"gencode.v41.annotation.gff3\")\n",
    "gff = annotation.df\n",
    "gff = gff.rename(columns={\"seq_id\":\"chr\"}) # 染色体の列名を chr に変更\n",
    "\n",
    "print(len(gff))\n",
    "\n",
    "gff.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "4a2ac361",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff[\"gene_type\"] = gff[\"attributes\"].apply(get_gene_type)\n",
    "gff = gff[gff[\"gene_type\"] == \"lncRNA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "0e3afb9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000243485.5;gene_id=ENSG00000243485.5...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29554</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000473358.1;Parent=ENSG00000243485.5;...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>29554</td>\n",
       "      <td>30039</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000473358.1:1;Parent=ENST0000047...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>30564</td>\n",
       "      <td>30667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000473358.1:2;Parent=ENST0000047...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>30976</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000473358.1:3;Parent=ENST0000047...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3373355</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57208519</td>\n",
       "      <td>57208756</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000483543.7_PAR_Y:4;Parent=ENST0...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3373356</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>57201143</td>\n",
       "      <td>57203357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000185203.12_PAR_Y;gene_id=ENSG000001...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3373357</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57201143</td>\n",
       "      <td>57203357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000399966.9_PAR_Y;Parent=ENSG00000185...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3373358</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57203182</td>\n",
       "      <td>57203357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000399966.9_PAR_Y:1;Parent=ENST0...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3373359</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57201143</td>\n",
       "      <td>57202020</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000399966.9_PAR_Y:2;Parent=ENST0...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>268144 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr  source        type     start       end score strand phase  \\\n",
       "28       chr1  HAVANA        gene     29554     31109     .      +     .   \n",
       "29       chr1  HAVANA  transcript     29554     31097     .      +     .   \n",
       "30       chr1  HAVANA        exon     29554     30039     .      +     .   \n",
       "31       chr1  HAVANA        exon     30564     30667     .      +     .   \n",
       "32       chr1  HAVANA        exon     30976     31097     .      +     .   \n",
       "...       ...     ...         ...       ...       ...   ...    ...   ...   \n",
       "3373355  chrY  HAVANA        exon  57208519  57208756     .      +     .   \n",
       "3373356  chrY  HAVANA        gene  57201143  57203357     .      -     .   \n",
       "3373357  chrY  HAVANA  transcript  57201143  57203357     .      -     .   \n",
       "3373358  chrY  HAVANA        exon  57203182  57203357     .      -     .   \n",
       "3373359  chrY  HAVANA        exon  57201143  57202020     .      -     .   \n",
       "\n",
       "                                                attributes gene_type  \n",
       "28       ID=ENSG00000243485.5;gene_id=ENSG00000243485.5...    lncRNA  \n",
       "29       ID=ENST00000473358.1;Parent=ENSG00000243485.5;...    lncRNA  \n",
       "30       ID=exon:ENST00000473358.1:1;Parent=ENST0000047...    lncRNA  \n",
       "31       ID=exon:ENST00000473358.1:2;Parent=ENST0000047...    lncRNA  \n",
       "32       ID=exon:ENST00000473358.1:3;Parent=ENST0000047...    lncRNA  \n",
       "...                                                    ...       ...  \n",
       "3373355  ID=exon:ENST00000483543.7_PAR_Y:4;Parent=ENST0...    lncRNA  \n",
       "3373356  ID=ENSG00000185203.12_PAR_Y;gene_id=ENSG000001...    lncRNA  \n",
       "3373357  ID=ENST00000399966.9_PAR_Y;Parent=ENSG00000185...    lncRNA  \n",
       "3373358  ID=exon:ENST00000399966.9_PAR_Y:1;Parent=ENST0...    lncRNA  \n",
       "3373359  ID=exon:ENST00000399966.9_PAR_Y:2;Parent=ENST0...    lncRNA  \n",
       "\n",
       "[268144 rows x 10 columns]"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4141630",
   "metadata": {},
   "source": [
    "`Ensembl_canonical` かつ `type` == `transcript` のものを抽出。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "9f2fcaf8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29554</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000473358.1;Parent=ENSG00000243485.5;...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000417324.1;Parent=ENSG00000237613.2;...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>92230</td>\n",
       "      <td>129217</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000477740.5;Parent=ENSG00000238009.6;...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>89551</td>\n",
       "      <td>91105</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000495576.1;Parent=ENSG00000239945.1;...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>139790</td>\n",
       "      <td>140339</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000493797.1;Parent=ENSG00000239906.1;...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3372838</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>25378300</td>\n",
       "      <td>25394719</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000427373.5;Parent=ENSG00000228786.5;...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3372887</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>25482908</td>\n",
       "      <td>25486705</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000306641.1;Parent=ENSG00000240450.1;...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3372986</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>25728490</td>\n",
       "      <td>25733388</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000417334.1;Parent=ENSG00000231141.1;...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3373351</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57190738</td>\n",
       "      <td>57208756</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000483543.7_PAR_Y;Parent=ENSG00000270...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3373357</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57201143</td>\n",
       "      <td>57203357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000399966.9_PAR_Y;Parent=ENSG00000185...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18041 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr  source        type     start       end score strand phase  \\\n",
       "29       chr1  HAVANA  transcript     29554     31097     .      +     .   \n",
       "40       chr1  HAVANA  transcript     34554     36081     .      -     .   \n",
       "75       chr1  HAVANA  transcript     92230    129217     .      -     .   \n",
       "93       chr1  HAVANA  transcript     89551     91105     .      -     .   \n",
       "106      chr1  HAVANA  transcript    139790    140339     .      -     .   \n",
       "...       ...     ...         ...       ...       ...   ...    ...   ...   \n",
       "3372838  chrY  HAVANA  transcript  25378300  25394719     .      -     .   \n",
       "3372887  chrY  HAVANA  transcript  25482908  25486705     .      +     .   \n",
       "3372986  chrY  HAVANA  transcript  25728490  25733388     .      +     .   \n",
       "3373351  chrY  HAVANA  transcript  57190738  57208756     .      +     .   \n",
       "3373357  chrY  HAVANA  transcript  57201143  57203357     .      -     .   \n",
       "\n",
       "                                                attributes gene_type  \n",
       "29       ID=ENST00000473358.1;Parent=ENSG00000243485.5;...    lncRNA  \n",
       "40       ID=ENST00000417324.1;Parent=ENSG00000237613.2;...    lncRNA  \n",
       "75       ID=ENST00000477740.5;Parent=ENSG00000238009.6;...    lncRNA  \n",
       "93       ID=ENST00000495576.1;Parent=ENSG00000239945.1;...    lncRNA  \n",
       "106      ID=ENST00000493797.1;Parent=ENSG00000239906.1;...    lncRNA  \n",
       "...                                                    ...       ...  \n",
       "3372838  ID=ENST00000427373.5;Parent=ENSG00000228786.5;...    lncRNA  \n",
       "3372887  ID=ENST00000306641.1;Parent=ENSG00000240450.1;...    lncRNA  \n",
       "3372986  ID=ENST00000417334.1;Parent=ENSG00000231141.1;...    lncRNA  \n",
       "3373351  ID=ENST00000483543.7_PAR_Y;Parent=ENSG00000270...    lncRNA  \n",
       "3373357  ID=ENST00000399966.9_PAR_Y;Parent=ENSG00000185...    lncRNA  \n",
       "\n",
       "[18041 rows x 10 columns]"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff = gff[gff[\"attributes\"].str.contains(\"Ensembl_canonical\")]\n",
    "gff = gff[gff[\"type\"] == \"transcript\"]\n",
    "gff"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83199844",
   "metadata": {},
   "source": [
    "遺伝子名, 遺伝子ID を抽出 (lncRNAであることもわかるようにする)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "1ead3b36",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_attributes_lncRNA(df):\n",
    "    attributes = df.split(\";\")\n",
    "    gene_name = [s for s in attributes if \"gene_name=\" in s][0].replace(\"gene_name=\", \"\")\n",
    "    gene_id = [s for s in attributes if \"gene_id=\" in s][0].replace(\"gene_id=\", \"\")\n",
    "    return gene_name + \"|\" + gene_id + \"|lncRNA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "92381274",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff[\"gene\"] = gff[\"attributes\"].apply(get_attributes_lncRNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "68f8d418",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29554</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000473358.1;Parent=ENSG00000243485.5;...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG|ENSG00000243485.5|lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000417324.1;Parent=ENSG00000237613.2;...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>FAM138A|ENSG00000237613.2|lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>92230</td>\n",
       "      <td>129217</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000477740.5;Parent=ENSG00000238009.6;...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENSG00000238009|ENSG00000238009.6|lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>89551</td>\n",
       "      <td>91105</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000495576.1;Parent=ENSG00000239945.1;...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENSG00000239945|ENSG00000239945.1|lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>139790</td>\n",
       "      <td>140339</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000493797.1;Parent=ENSG00000239906.1;...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENSG00000239906|ENSG00000239906.1|lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3372838</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>25378300</td>\n",
       "      <td>25394719</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000427373.5;Parent=ENSG00000228786.5;...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>SEPTIN14P23|ENSG00000228786.5|lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3372887</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>25482908</td>\n",
       "      <td>25486705</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000306641.1;Parent=ENSG00000240450.1;...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>CSPG4P1Y|ENSG00000240450.1|lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3372986</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>25728490</td>\n",
       "      <td>25733388</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000417334.1;Parent=ENSG00000231141.1;...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TTTY3|ENSG00000231141.1|lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3373351</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57190738</td>\n",
       "      <td>57208756</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000483543.7_PAR_Y;Parent=ENSG00000270...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ENSG00000270726|ENSG00000270726.6|lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3373357</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57201143</td>\n",
       "      <td>57203357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000399966.9_PAR_Y;Parent=ENSG00000185...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>WASIR1|ENSG00000185203.12|lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18041 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr  source        type     start       end score strand phase  \\\n",
       "29       chr1  HAVANA  transcript     29554     31097     .      +     .   \n",
       "40       chr1  HAVANA  transcript     34554     36081     .      -     .   \n",
       "75       chr1  HAVANA  transcript     92230    129217     .      -     .   \n",
       "93       chr1  HAVANA  transcript     89551     91105     .      -     .   \n",
       "106      chr1  HAVANA  transcript    139790    140339     .      -     .   \n",
       "...       ...     ...         ...       ...       ...   ...    ...   ...   \n",
       "3372838  chrY  HAVANA  transcript  25378300  25394719     .      -     .   \n",
       "3372887  chrY  HAVANA  transcript  25482908  25486705     .      +     .   \n",
       "3372986  chrY  HAVANA  transcript  25728490  25733388     .      +     .   \n",
       "3373351  chrY  HAVANA  transcript  57190738  57208756     .      +     .   \n",
       "3373357  chrY  HAVANA  transcript  57201143  57203357     .      -     .   \n",
       "\n",
       "                                                attributes gene_type  \\\n",
       "29       ID=ENST00000473358.1;Parent=ENSG00000243485.5;...    lncRNA   \n",
       "40       ID=ENST00000417324.1;Parent=ENSG00000237613.2;...    lncRNA   \n",
       "75       ID=ENST00000477740.5;Parent=ENSG00000238009.6;...    lncRNA   \n",
       "93       ID=ENST00000495576.1;Parent=ENSG00000239945.1;...    lncRNA   \n",
       "106      ID=ENST00000493797.1;Parent=ENSG00000239906.1;...    lncRNA   \n",
       "...                                                    ...       ...   \n",
       "3372838  ID=ENST00000427373.5;Parent=ENSG00000228786.5;...    lncRNA   \n",
       "3372887  ID=ENST00000306641.1;Parent=ENSG00000240450.1;...    lncRNA   \n",
       "3372986  ID=ENST00000417334.1;Parent=ENSG00000231141.1;...    lncRNA   \n",
       "3373351  ID=ENST00000483543.7_PAR_Y;Parent=ENSG00000270...    lncRNA   \n",
       "3373357  ID=ENST00000399966.9_PAR_Y;Parent=ENSG00000185...    lncRNA   \n",
       "\n",
       "                                             gene  \n",
       "29           MIR1302-2HG|ENSG00000243485.5|lncRNA  \n",
       "40               FAM138A|ENSG00000237613.2|lncRNA  \n",
       "75       ENSG00000238009|ENSG00000238009.6|lncRNA  \n",
       "93       ENSG00000239945|ENSG00000239945.1|lncRNA  \n",
       "106      ENSG00000239906|ENSG00000239906.1|lncRNA  \n",
       "...                                           ...  \n",
       "3372838      SEPTIN14P23|ENSG00000228786.5|lncRNA  \n",
       "3372887         CSPG4P1Y|ENSG00000240450.1|lncRNA  \n",
       "3372986            TTTY3|ENSG00000231141.1|lncRNA  \n",
       "3373351  ENSG00000270726|ENSG00000270726.6|lncRNA  \n",
       "3373357          WASIR1|ENSG00000185203.12|lncRNA  \n",
       "\n",
       "[18041 rows x 11 columns]"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "d7b919cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 18041\n",
      "Unique rows: 18027\n"
     ]
    }
   ],
   "source": [
    "rows = len(gff)\n",
    "unique_rows = len(gff[\"gene\"].unique())\n",
    "\n",
    "print(f\"Number of rows: {rows}\")\n",
    "print(f\"Unique rows: {unique_rows}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcc51057",
   "metadata": {},
   "source": [
    "lncRNAも重複があるものがあるが、これは性染色体（X, Y）に同じ遺伝子を持っているため。なので今回は無視。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "8ccf2de6",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff = gff[[\"chr\", \"start\", \"end\", \"gene\", \"score\", \"strand\"]].drop_duplicates().reset_index(drop=True)\n",
    "gff[\"score\"] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "2e7a1657",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31097</td>\n",
       "      <td>MIR1302-2HG|ENSG00000243485.5|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>FAM138A|ENSG00000237613.2|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92230</td>\n",
       "      <td>129217</td>\n",
       "      <td>ENSG00000238009|ENSG00000238009.6|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>89551</td>\n",
       "      <td>91105</td>\n",
       "      <td>ENSG00000239945|ENSG00000239945.1|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>139790</td>\n",
       "      <td>140339</td>\n",
       "      <td>ENSG00000239906|ENSG00000239906.1|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18036</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25378300</td>\n",
       "      <td>25394719</td>\n",
       "      <td>SEPTIN14P23|ENSG00000228786.5|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18037</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25482908</td>\n",
       "      <td>25486705</td>\n",
       "      <td>CSPG4P1Y|ENSG00000240450.1|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18038</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25728490</td>\n",
       "      <td>25733388</td>\n",
       "      <td>TTTY3|ENSG00000231141.1|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18039</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57190738</td>\n",
       "      <td>57208756</td>\n",
       "      <td>ENSG00000270726|ENSG00000270726.6|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18040</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201143</td>\n",
       "      <td>57203357</td>\n",
       "      <td>WASIR1|ENSG00000185203.12|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18041 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr     start       end                                      gene  \\\n",
       "0      chr1     29554     31097      MIR1302-2HG|ENSG00000243485.5|lncRNA   \n",
       "1      chr1     34554     36081          FAM138A|ENSG00000237613.2|lncRNA   \n",
       "2      chr1     92230    129217  ENSG00000238009|ENSG00000238009.6|lncRNA   \n",
       "3      chr1     89551     91105  ENSG00000239945|ENSG00000239945.1|lncRNA   \n",
       "4      chr1    139790    140339  ENSG00000239906|ENSG00000239906.1|lncRNA   \n",
       "...     ...       ...       ...                                       ...   \n",
       "18036  chrY  25378300  25394719      SEPTIN14P23|ENSG00000228786.5|lncRNA   \n",
       "18037  chrY  25482908  25486705         CSPG4P1Y|ENSG00000240450.1|lncRNA   \n",
       "18038  chrY  25728490  25733388            TTTY3|ENSG00000231141.1|lncRNA   \n",
       "18039  chrY  57190738  57208756  ENSG00000270726|ENSG00000270726.6|lncRNA   \n",
       "18040  chrY  57201143  57203357          WASIR1|ENSG00000185203.12|lncRNA   \n",
       "\n",
       "       score strand  \n",
       "0          0      +  \n",
       "1          0      -  \n",
       "2          0      -  \n",
       "3          0      -  \n",
       "4          0      -  \n",
       "...      ...    ...  \n",
       "18036      0      -  \n",
       "18037      0      +  \n",
       "18038      0      +  \n",
       "18039      0      +  \n",
       "18040      0      -  \n",
       "\n",
       "[18041 rows x 6 columns]"
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5656f02",
   "metadata": {},
   "source": [
    "同じ染色体で重複のある遺伝子について 3' UTR の開始位置の最小値と終了位置の最大値を取得しまとめる。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "79edaa95",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_unique = gff[gff[\"gene\"].duplicated(keep=False) == False]\n",
    "gff_dup = gff[gff[\"gene\"].duplicated(keep=False)]\n",
    "\n",
    "duplicated_genes = np.array(gff_dup[\"gene\"].unique())\n",
    "gff_dedup = pd.DataFrame(columns=[\"chr\", \"start\", \"end\", \"gene\", \"score\", \"strand\"])\n",
    "\n",
    "for gene in duplicated_genes:\n",
    "    tmp_df = gff_dup[gff_dup[\"gene\"] == gene]   \n",
    "    tmp_df[\"chr\"].unique()   \n",
    "    \n",
    "    if len(tmp_df[\"chr\"].unique()) == 1:\n",
    "        tmp_chr = tmp_df.iloc[0, 0]\n",
    "        tmp_start = tmp_df[\"start\"].min()\n",
    "        tmp_end = tmp_df[\"end\"].max()\n",
    "        tmp_strand = tmp_df.iloc[0, 5]\n",
    "        add_row = pd.DataFrame([tmp_chr, tmp_start, tmp_end, gene, 0, tmp_strand], index=gff_dedup.columns).T\n",
    "        gff_dedup = pd.concat([gff_dedup, add_row])\n",
    "    \n",
    "    else:\n",
    "        gff_dedup = pd.concat([gff_dedup, tmp_df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "3c0aa3d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31097</td>\n",
       "      <td>MIR1302-2HG|ENSG00000243485.5|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>FAM138A|ENSG00000237613.2|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92230</td>\n",
       "      <td>129217</td>\n",
       "      <td>ENSG00000238009|ENSG00000238009.6|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>89551</td>\n",
       "      <td>91105</td>\n",
       "      <td>ENSG00000239945|ENSG00000239945.1|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>139790</td>\n",
       "      <td>140339</td>\n",
       "      <td>ENSG00000239906|ENSG00000239906.1|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17971</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2612988</td>\n",
       "      <td>2615347</td>\n",
       "      <td>LINC00102|ENSG00000230542.6|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17958</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156004218</td>\n",
       "      <td>156022236</td>\n",
       "      <td>ENSG00000270726|ENSG00000270726.6|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18039</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57190738</td>\n",
       "      <td>57208756</td>\n",
       "      <td>ENSG00000270726|ENSG00000270726.6|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17959</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156014623</td>\n",
       "      <td>156016837</td>\n",
       "      <td>WASIR1|ENSG00000185203.12|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18040</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201143</td>\n",
       "      <td>57203357</td>\n",
       "      <td>WASIR1|ENSG00000185203.12|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18041 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr      start        end                                      gene  \\\n",
       "0      chr1      29554      31097      MIR1302-2HG|ENSG00000243485.5|lncRNA   \n",
       "1      chr1      34554      36081          FAM138A|ENSG00000237613.2|lncRNA   \n",
       "2      chr1      92230     129217  ENSG00000238009|ENSG00000238009.6|lncRNA   \n",
       "3      chr1      89551      91105  ENSG00000239945|ENSG00000239945.1|lncRNA   \n",
       "4      chr1     139790     140339  ENSG00000239906|ENSG00000239906.1|lncRNA   \n",
       "...     ...        ...        ...                                       ...   \n",
       "17971  chrY    2612988    2615347        LINC00102|ENSG00000230542.6|lncRNA   \n",
       "17958  chrX  156004218  156022236  ENSG00000270726|ENSG00000270726.6|lncRNA   \n",
       "18039  chrY   57190738   57208756  ENSG00000270726|ENSG00000270726.6|lncRNA   \n",
       "17959  chrX  156014623  156016837          WASIR1|ENSG00000185203.12|lncRNA   \n",
       "18040  chrY   57201143   57203357          WASIR1|ENSG00000185203.12|lncRNA   \n",
       "\n",
       "      score strand  \n",
       "0         0      +  \n",
       "1         0      -  \n",
       "2         0      -  \n",
       "3         0      -  \n",
       "4         0      -  \n",
       "...     ...    ...  \n",
       "17971     0      -  \n",
       "17958     0      +  \n",
       "18039     0      +  \n",
       "17959     0      -  \n",
       "18040     0      -  \n",
       "\n",
       "[18041 rows x 6 columns]"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_merged_lncRNA = pd.concat([gff_unique, gff_dedup])\n",
    "gff_merged_lncRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "ca21c608",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>70009</td>\n",
       "      <td>71585</td>\n",
       "      <td>OR4F5|ENSG00000186092.7</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944154</td>\n",
       "      <td>944574</td>\n",
       "      <td>SAMD11|ENSG00000187634.13</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944203</td>\n",
       "      <td>944693</td>\n",
       "      <td>NOC2L|ENSG00000188976.11</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>965192</td>\n",
       "      <td>965719</td>\n",
       "      <td>KLHL17|ENSG00000187961.15</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>974576</td>\n",
       "      <td>975865</td>\n",
       "      <td>PLEKHN1|ENSG00000187583.11</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17971</th>\n",
       "      <td>chrY</td>\n",
       "      <td>2612988</td>\n",
       "      <td>2615347</td>\n",
       "      <td>LINC00102|ENSG00000230542.6|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17958</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156004218</td>\n",
       "      <td>156022236</td>\n",
       "      <td>ENSG00000270726|ENSG00000270726.6|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18039</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57190738</td>\n",
       "      <td>57208756</td>\n",
       "      <td>ENSG00000270726|ENSG00000270726.6|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17959</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156014623</td>\n",
       "      <td>156016837</td>\n",
       "      <td>WASIR1|ENSG00000185203.12|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18040</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201143</td>\n",
       "      <td>57203357</td>\n",
       "      <td>WASIR1|ENSG00000185203.12|lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37545 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr      start        end                                      gene  \\\n",
       "0      chr1      70009      71585                   OR4F5|ENSG00000186092.7   \n",
       "1      chr1     944154     944574                 SAMD11|ENSG00000187634.13   \n",
       "2      chr1     944203     944693                  NOC2L|ENSG00000188976.11   \n",
       "3      chr1     965192     965719                 KLHL17|ENSG00000187961.15   \n",
       "4      chr1     974576     975865                PLEKHN1|ENSG00000187583.11   \n",
       "...     ...        ...        ...                                       ...   \n",
       "17971  chrY    2612988    2615347        LINC00102|ENSG00000230542.6|lncRNA   \n",
       "17958  chrX  156004218  156022236  ENSG00000270726|ENSG00000270726.6|lncRNA   \n",
       "18039  chrY   57190738   57208756  ENSG00000270726|ENSG00000270726.6|lncRNA   \n",
       "17959  chrX  156014623  156016837          WASIR1|ENSG00000185203.12|lncRNA   \n",
       "18040  chrY   57201143   57203357          WASIR1|ENSG00000185203.12|lncRNA   \n",
       "\n",
       "      score strand  \n",
       "0         0      +  \n",
       "1         0      +  \n",
       "2         0      -  \n",
       "3         0      +  \n",
       "4         0      +  \n",
       "...     ...    ...  \n",
       "17971     0      -  \n",
       "17958     0      +  \n",
       "18039     0      +  \n",
       "17959     0      -  \n",
       "18040     0      -  \n",
       "\n",
       "[37545 rows x 6 columns]"
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_merged = pd.concat([gff_merged, gff_merged_lncRNA])\n",
    "gff_merged"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da11f929",
   "metadata": {},
   "source": [
    "`protein_coding`と`lncRNA`をあわせて37,524遺伝子の情報を抽出できた。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "a27b8e6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_merged.to_csv(\"gencode.v41.customized.canonical.with.lncRNA.bed\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3d31a1b",
   "metadata": {},
   "source": [
    "### Q. 3'UTRだけで本当にすべてのリードをカバーしきれているのか？\n",
    "マッピング後のBAMファイルをIGVで見ると、3'UTRだけでなくラストエクソン（またはさらにその1つ前のエクソン）にまでまたがってリードがマッピングされている遺伝子がいくつも見つかる。そういう遺伝子では3'UTRに貼り付いたリードのみをカウントすると、リード数が少なめに見積もられる。これが積み重なると`DESeq2`で発現変動解析を行う際に size factor に影響が出るのでは？"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b430dbe1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
