{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0fe16e98",
   "metadata": {},
   "source": [
    "## gencode.v41.annotation.gff3 の読み込み\n",
    "gff[\"attributes\"]に様々な情報が格納されている。各情報は\";\"で区切られているので`split(\";\")`で情報を抽出できる。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f9bc9321",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gffpandas.gffpandas as gffpd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6a4581c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3373604\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000223972.5;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  source        type  start    end score strand phase  \\\n",
       "0  chr1  HAVANA        gene  11869  14409     .      +     .   \n",
       "1  chr1  HAVANA  transcript  11869  14409     .      +     .   \n",
       "2  chr1  HAVANA        exon  11869  12227     .      +     .   \n",
       "3  chr1  HAVANA        exon  12613  12721     .      +     .   \n",
       "4  chr1  HAVANA        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...  \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000223972.5;...  \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation = gffpd.read_gff3(\"gencode.v41.annotation.gff3\")\n",
    "gff = annotation.df\n",
    "gff = gff.rename(columns={\"seq_id\":\"chr\"}) # 染色体の列名を chr に変更\n",
    "\n",
    "print(len(gff))\n",
    "\n",
    "gff.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82e9951a",
   "metadata": {},
   "source": [
    "タンパク質をコードする遺伝子（`gene_type=\"protein_coding\"`）については、GFFファイル中に3'UTRの情報(`type=\"three_prime_RNA\"`)が含まれており、その情報を抽出すればOK。一方、lncRNA（`gene_type=lncRNA`）については、GFFファイル中に3'UTRの情報が含まれていないため、代わりに遺伝子全長の情報（`type=\"transcript\"`）を抽出する。GFFファイルには1つの遺伝子あたり複数の転写産物が登録されているが、その中で最も代表的なもの (`Ensembl_canonical`) がEnsemblのアルゴリズムによって選出されている。この**canonical**な転写産物のみを抽出する。\n",
    "### Ensembl Canonical transcript\n",
    "The Ensembl Canonical transcript is a single, representative transcript identified at every locus. For accurate analysis, we recommend that more than one transcripts at a locus may need to be considered, however, we designate a single Ensembl Canonical transcript per locus to provide consistency when only one transcript is required e.g. for initial display on the Ensembl (or other) website and for use in genome-wide calculations e.g. the Ensembl gene tree analysis.\n",
    "\n",
    "For protein-coding genes, we aim to identify the transcript that, on balance, has the highest coverage of conserved exons, highest expression, longest coding sequence and is represented in other key resources, such as NCBI and UniProt. To identify this transcript, we consider, where available, evidence of functional potential (such as evolutionary conservation of a coding region, transcript expression levels), transcript length and evidence from other resources (such as concordance with the APPRIS1 ’principal isoform’ and with the UniProt/Swiss-Prot ‘canonical isoform’)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80ac72cc",
   "metadata": {},
   "source": [
    "まず`gene_type`を抽出する関数を用意。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "966e16ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_type(df):\n",
    "    attributes = df.split(\";\")\n",
    "    output_type = [s for s in attributes if \"gene_type=\" in s][0].replace(\"gene_type=\", \"\")\n",
    "    return output_type"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "973b7abf",
   "metadata": {},
   "source": [
    "### (1) タンパク質をコードしている遺伝子のみを抽出\n",
    "`gene_type` が `protein_coding` かつ `Ensembl_canonical` なものを抽出。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "744bd93f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff[\"gene_type\"] = gff[\"attributes\"].apply(get_gene_type)\n",
    "gff = gff[gff[\"gene_type\"] == \"protein_coding\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c51cd0ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 2988184\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000186092.7;gene_id=ENSG00000186092.7...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000641515.2;Parent=ENSG00000186092.7;...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65419</td>\n",
       "      <td>65433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000641515.2:1;Parent=ENST0000064...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65520</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000641515.2:2;Parent=ENST0000064...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65565</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENST00000641515.2;Parent=ENST0000064151...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr  source        type  start    end score strand phase  \\\n",
       "57  chr1  HAVANA        gene  65419  71585     .      +     .   \n",
       "58  chr1  HAVANA  transcript  65419  71585     .      +     .   \n",
       "59  chr1  HAVANA        exon  65419  65433     .      +     .   \n",
       "60  chr1  HAVANA        exon  65520  65573     .      +     .   \n",
       "61  chr1  HAVANA         CDS  65565  65573     .      +     0   \n",
       "\n",
       "                                           attributes       gene_type  \n",
       "57  ID=ENSG00000186092.7;gene_id=ENSG00000186092.7...  protein_coding  \n",
       "58  ID=ENST00000641515.2;Parent=ENSG00000186092.7;...  protein_coding  \n",
       "59  ID=exon:ENST00000641515.2:1;Parent=ENST0000064...  protein_coding  \n",
       "60  ID=exon:ENST00000641515.2:2;Parent=ENST0000064...  protein_coding  \n",
       "61  ID=CDS:ENST00000641515.2;Parent=ENST0000064151...  protein_coding  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Number of rows: {len(gff)}\")\n",
    "gff.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a4e34c12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['gene', 'transcript', 'exon', 'CDS', 'start_codon', 'stop_codon',\n",
       "       'five_prime_UTR', 'three_prime_UTR',\n",
       "       'stop_codon_redefined_as_selenocysteine'], dtype=object)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff[\"type\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5831d1fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 518563\n"
     ]
    }
   ],
   "source": [
    "gff = gff[gff[\"attributes\"].str.contains(\"Ensembl_canonical\")]\n",
    "\n",
    "print(f\"Number of rows: {len(gff)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bf6cec7",
   "metadata": {},
   "source": [
    "Coding sequence (type=`CDS`) の情報のみを抽出する。Sourceは `HAVANA` を利用。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "214c75cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 197279\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65565</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENST00000641515.2;Parent=ENST0000064151...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69037</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENST00000641515.2;Parent=ENST0000064151...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENST00000426406.4;Parent=ENST0000042640...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENST00000332831.5;Parent=ENST0000033283...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>978</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>924432</td>\n",
       "      <td>924948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENST00000616016.5;Parent=ENST0000061601...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr  source type   start     end score strand phase  \\\n",
       "61   chr1  HAVANA  CDS   65565   65573     .      +     0   \n",
       "64   chr1  HAVANA  CDS   69037   70008     .      +     0   \n",
       "322  chr1  HAVANA  CDS  450740  451678     .      -     0   \n",
       "448  chr1  HAVANA  CDS  685716  686654     .      -     0   \n",
       "978  chr1  HAVANA  CDS  924432  924948     .      +     0   \n",
       "\n",
       "                                            attributes       gene_type  \n",
       "61   ID=CDS:ENST00000641515.2;Parent=ENST0000064151...  protein_coding  \n",
       "64   ID=CDS:ENST00000641515.2;Parent=ENST0000064151...  protein_coding  \n",
       "322  ID=CDS:ENST00000426406.4;Parent=ENST0000042640...  protein_coding  \n",
       "448  ID=CDS:ENST00000332831.5;Parent=ENST0000033283...  protein_coding  \n",
       "978  ID=CDS:ENST00000616016.5;Parent=ENST0000061601...  protein_coding  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff = gff[(gff[\"type\"] == \"CDS\") & (gff[\"source\"] == \"HAVANA\")]\n",
    "print(f\"Number of rows: {len(gff)}\")\n",
    "gff.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0656b332",
   "metadata": {},
   "source": [
    "BEDファイルに記載しておく情報として `gene_name` と `gene_id` の2つの情報があればいい。これらの情報を抽出しする関数を作成。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "75f28a9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_attributes(df):\n",
    "    attributes = df.split(\";\")\n",
    "    gene_name = [s for s in attributes if \"gene_name=\" in s][0].replace(\"gene_name=\", \"\")\n",
    "    gene_id = [s for s in attributes if \"gene_id=\" in s][0].replace(\"gene_id=\", \"\")\n",
    "    return gene_name + \"|\" + gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "36be40ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff[\"gene\"] = gff[\"attributes\"].apply(get_attributes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "fe38f2d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 197279\n",
      "Unique rows: 19974\n"
     ]
    }
   ],
   "source": [
    "rows = len(gff)\n",
    "unique_rows = len(gff[\"gene\"].unique())\n",
    "\n",
    "print(f\"Number of rows: {rows}\")\n",
    "print(f\"Unique rows: {unique_rows}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0563bc64",
   "metadata": {},
   "source": [
    "ユニークな遺伝子名が19,974個なのに行数が多い。CDS がエクソンとして分割されているため。CDS　の長さを計算後、それらを足し合わせた数字を ORF の長さとする。"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72219718",
   "metadata": {},
   "source": [
    "まずBEDファイルに必要な情報 (`chr`, `start`, `end`, `gene`, `score`, `strand`) を抽出"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "7eea5a0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene</th>\n",
       "      <th>length</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65565</td>\n",
       "      <td>65573</td>\n",
       "      <td>OR4F5|ENSG00000186092.7</td>\n",
       "      <td>8</td>\n",
       "      <td>OR4F5|ENSG00000186092.7|chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69037</td>\n",
       "      <td>70008</td>\n",
       "      <td>OR4F5|ENSG00000186092.7</td>\n",
       "      <td>971</td>\n",
       "      <td>OR4F5|ENSG00000186092.7|chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>OR4F29|ENSG00000284733.2</td>\n",
       "      <td>938</td>\n",
       "      <td>OR4F29|ENSG00000284733.2|chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>OR4F16|ENSG00000284662.2</td>\n",
       "      <td>938</td>\n",
       "      <td>OR4F16|ENSG00000284662.2|chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>924432</td>\n",
       "      <td>924948</td>\n",
       "      <td>SAMD11|ENSG00000187634.13</td>\n",
       "      <td>516</td>\n",
       "      <td>SAMD11|ENSG00000187634.13|chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197274</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209532</td>\n",
       "      <td>57209733</td>\n",
       "      <td>WASH6P|ENSG00000182484.15</td>\n",
       "      <td>201</td>\n",
       "      <td>WASH6P|ENSG00000182484.15|chrY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197275</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57209822</td>\n",
       "      <td>57209980</td>\n",
       "      <td>WASH6P|ENSG00000182484.15</td>\n",
       "      <td>158</td>\n",
       "      <td>WASH6P|ENSG00000182484.15|chrY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197276</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57210640</td>\n",
       "      <td>57210792</td>\n",
       "      <td>WASH6P|ENSG00000182484.15</td>\n",
       "      <td>152</td>\n",
       "      <td>WASH6P|ENSG00000182484.15|chrY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197277</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57211552</td>\n",
       "      <td>57211620</td>\n",
       "      <td>WASH6P|ENSG00000182484.15</td>\n",
       "      <td>68</td>\n",
       "      <td>WASH6P|ENSG00000182484.15|chrY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197278</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57211761</td>\n",
       "      <td>57211894</td>\n",
       "      <td>WASH6P|ENSG00000182484.15</td>\n",
       "      <td>133</td>\n",
       "      <td>WASH6P|ENSG00000182484.15|chrY</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>197279 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr     start       end                       gene  length  \\\n",
       "0       chr1     65565     65573    OR4F5|ENSG00000186092.7       8   \n",
       "1       chr1     69037     70008    OR4F5|ENSG00000186092.7     971   \n",
       "2       chr1    450740    451678   OR4F29|ENSG00000284733.2     938   \n",
       "3       chr1    685716    686654   OR4F16|ENSG00000284662.2     938   \n",
       "4       chr1    924432    924948  SAMD11|ENSG00000187634.13     516   \n",
       "...      ...       ...       ...                        ...     ...   \n",
       "197274  chrY  57209532  57209733  WASH6P|ENSG00000182484.15     201   \n",
       "197275  chrY  57209822  57209980  WASH6P|ENSG00000182484.15     158   \n",
       "197276  chrY  57210640  57210792  WASH6P|ENSG00000182484.15     152   \n",
       "197277  chrY  57211552  57211620  WASH6P|ENSG00000182484.15      68   \n",
       "197278  chrY  57211761  57211894  WASH6P|ENSG00000182484.15     133   \n",
       "\n",
       "                                    ID  \n",
       "0         OR4F5|ENSG00000186092.7|chr1  \n",
       "1         OR4F5|ENSG00000186092.7|chr1  \n",
       "2        OR4F29|ENSG00000284733.2|chr1  \n",
       "3        OR4F16|ENSG00000284662.2|chr1  \n",
       "4       SAMD11|ENSG00000187634.13|chr1  \n",
       "...                                ...  \n",
       "197274  WASH6P|ENSG00000182484.15|chrY  \n",
       "197275  WASH6P|ENSG00000182484.15|chrY  \n",
       "197276  WASH6P|ENSG00000182484.15|chrY  \n",
       "197277  WASH6P|ENSG00000182484.15|chrY  \n",
       "197278  WASH6P|ENSG00000182484.15|chrY  \n",
       "\n",
       "[197279 rows x 6 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff = gff[[\"chr\", \"start\", \"end\", \"gene\"]].drop_duplicates().reset_index(drop=True)\n",
    "gff[\"length\"] = abs(gff[\"start\"] - gff[\"end\"])\n",
    "gff[\"ID\"] = gff[\"gene\"] + \"|\" + gff[\"chr\"]\n",
    "gff"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "822e4a3d",
   "metadata": {},
   "source": [
    "同じ染色体で重複のある遺伝子について 3' UTR の開始位置の最小値と終了位置の最大値を取得しまとめる。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "4456dbf4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OR4F5|ENSG00000186092.7|chr1</td>\n",
       "      <td>981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OR4F29|ENSG00000284733.2|chr1</td>\n",
       "      <td>939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OR4F16|ENSG00000284662.2|chr1</td>\n",
       "      <td>939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMD11|ENSG00000187634.13|chr1</td>\n",
       "      <td>2535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NOC2L|ENSG00000188976.11|chr1</td>\n",
       "      <td>2250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BPY2C|ENSG00000185894.8|chrY</td>\n",
       "      <td>321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDY1|ENSG00000172288.8|chrY</td>\n",
       "      <td>1665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VAMP7|ENSG00000124333.16|chrY</td>\n",
       "      <td>663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IL9R|ENSG00000124334.18|chrY</td>\n",
       "      <td>1566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WASH6P|ENSG00000182484.15|chrY</td>\n",
       "      <td>1437</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19992 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                ID length\n",
       "0     OR4F5|ENSG00000186092.7|chr1    981\n",
       "0    OR4F29|ENSG00000284733.2|chr1    939\n",
       "0    OR4F16|ENSG00000284662.2|chr1    939\n",
       "0   SAMD11|ENSG00000187634.13|chr1   2535\n",
       "0    NOC2L|ENSG00000188976.11|chr1   2250\n",
       "..                             ...    ...\n",
       "0     BPY2C|ENSG00000185894.8|chrY    321\n",
       "0      CDY1|ENSG00000172288.8|chrY   1665\n",
       "0    VAMP7|ENSG00000124333.16|chrY    663\n",
       "0     IL9R|ENSG00000124334.18|chrY   1566\n",
       "0   WASH6P|ENSG00000182484.15|chrY   1437\n",
       "\n",
       "[19992 rows x 2 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_ID = np.array(gff[\"ID\"].unique())\n",
    "output = pd.DataFrame(columns=[\"ID\", \"length\"])\n",
    "\n",
    "for i in gene_ID:\n",
    "    tmp_df = gff[gff[\"ID\"] == i]\n",
    "    tmp_ORF = tmp_df[\"length\"].sum() + len(tmp_df)\n",
    "    add_row = pd.DataFrame([i, tmp_ORF], index=[\"ID\", \"length\"]).T\n",
    "    output = pd.concat([output, add_row])\n",
    "\n",
    "output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2880e2b6",
   "metadata": {},
   "source": [
    "BEDファイルを保存。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "537d98b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "output.to_csv(\"gencode.v41.orf.length.tsv\", sep=\"\\t\", index=False, header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3d31a1b",
   "metadata": {},
   "source": [
    "### Q. 3'UTRだけで本当にすべてのリードをカバーしきれているのか？\n",
    "マッピング後のBAMファイルをIGVで見ると、3'UTRだけでなくラストエクソン（またはさらにその1つ前のエクソン）にまでまたがってリードがマッピングされている遺伝子がいくつも見つかる。そういう遺伝子では3'UTRに貼り付いたリードのみをカウントすると、リード数が少なめに見積もられる。これが積み重なると`DESeq2`で発現変動解析を行う際に size factor に影響が出るのでは？"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b430dbe1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
